File: 2794-maximum-number-of-moves-in-a-grid/maximum-number-of-moves-in-a-grid.java
class Solution {

    private int n, m;
    private int[][] dp;
    private int solve(int row, int col, int prev, int[][] grid)
    {
        if(row<0 || row >= n || col < 0 || col >= m || grid[row][col] <= prev)
        return 0;

        if(dp[row][col] != -1)
        return dp[row][col];

        int moves;

        int path1 = 1 + solve(row - 1, col + 1, grid[row][col], grid);
        int path2 = 1 + solve(row, col + 1, grid[row][col], grid);
        int path3 = 1 + solve(row + 1, col + 1, grid[row][col], grid);

        moves = Math.max(path1, Math.max(path2, path3));
        return dp[row][col] = moves;
    }

    public int maxMoves(int[][] grid) 
    {
        n =grid.length;
        m =grid[0].length;

        dp =  new int [n][m];
        for(int i = 0; i < n; i++)
        Arrays.fill(dp[i],-1);

        int moves = 0;
        for(int i = 0; i < n; i++)
        moves = Math.max(moves, solve(i,0,0,grid));

        return moves - 1;
    }
}

File: 2794-maximum-number-of-moves-in-a-grid/maximum-number-of-moves-in-a-grid.cpp
class Solution {
public:
    int M, N;
    // The three possible directions for the next column.
    const int dirs[3] = {-1, 0, 1};
    bool isValidCell(int row, int col)
    {
        return row >= 0 && col >= 0 && row < M && col < N;
    }
    int maxMoves(vector<vector<int>>& grid) 
    {
        M = grid.size();
        N = grid[0].size();

        queue<pair<int, int>> q;
        vector<vector<int>> vis(M, vector<int>(N, 0));
        // Enqueue the cells in the first column.
        for (int i = 0; i < M; i++) 
        {
            vis[i][0] = 1;
            q.push({i, 0});
        }

        int moves = 0;
        while(!q.empty()) 
        {
            int sz = q.size();
            while(sz--) 
            {
                auto [row, col] = q.front();
                q.pop();

                for (int dir : dirs) 
                {
                    // Next cell after the move.
                    int newRow = row + dir, newCol = col + 1;

                    // If the next cell isn't visited yet and is greater than
                    // the current cell value. Add it to the queue with the
                    // moves required.
                    if(isValidCell(newRow, newCol) && !vis[newRow][newCol] && grid[row][col] < grid[newRow][newCol]) 
                    {
                        vis[newRow][newCol] = 1;
                        q.push({newRow, newCol});
                    }
                }
            }
            moves++;
        }
        return moves - 1;
    }
};